### PR TITLE
Implement an additional check in the reducer for setting the debtee to avoid information loss

### LINKED ISSUE #13: User added through the "Who paid?" dialog disappears once another user is selected
To reproduce:

- Navigate to add bill screen
- Open "Who paid?" dialog
- Add a new user
- Open the dialog again
- Pick a different user

__Expected__ :
The new user is selected and the previously added user is still available

__Actual__ :
The new user is selected but the previously added one disappears

### PR DESCRIPTION
**Pull request by thomaseizinger**
Fixes #13.

Initially, we relied on a property that no longer exists on the action. (`isNewMember`).
Therefore, this was always false and the debtee (if added via the input box), was never explicitly added.

By adding a check to the reducer function, we add new persons explicitly to the bill through a "PersonValue".

### CODE REVIEW

File: src/routes/App/routes/new/modules/addBillForm/reducer.js [hunk 1]
```diff
@@ -29,29 +42,39 @@ export const addBillFormReducer = (state = initialState, action) => {
...
+      let {debtee} = action.payload;
 
-      if (!debtee && !isNewMember) {
+      if (!debtee || debtee === state.debtee) {
         return state;
       }
 
-      if (debtee === state.debtee) {
-        return state;
+      if (isTransientMember(debtee)) {
```
Comment (duffleit) [inline src/routes/App/routes/new/modules/addBillForm/reducer.js]: calling a method and writing a comment about the return value of this method seems like a code smell IMHO. Maybe just rename the function to `isMemeberInCurrentBill`?

File: src/routes/App/routes/new/modules/addBillForm/reducer.js [hunk 2]
```diff
@@ -29,29 +42,39 @@ export const addBillFormReducer = (state = initialState, action) => {
...
+        ]
+      };
     }
 
     case "NewDebteeSelected": {
-      let {debtee, isNewMember} = action.payload;
+      let {debtee} = action.payload;
 
-      if (!debtee && !isNewMember) {
+      if (!debtee || debtee === state.debtee) {
```
Comment (duffleit) [inline src/routes/App/routes/new/modules/addBillForm/reducer.js]: how about extracting variables? Something like `debteeNotSet` and `debteeNotChanged`

File: src/routes/App/routes/new/modules/addBillForm/reducer.js [hunk 3]
```diff
@@ -29,29 +42,39 @@ export const addBillFormReducer = (state = initialState, action) => {
...
-
-      return stateCopy
+      // Member is not yet in the current bill. Explicitly add them and return the new state.
+      return {
+        ...state,
+        debtee: debtee,
+        personValues: [
+          ...state.personValues,
+          createPersonValue(debtee)
+        ]
```
Comment (duffleit) [inline src/routes/App/routes/new/modules/addBillForm/reducer.js]: I would aggregate the personvalues in a single function and not within in the return statement, as it is duplicated in the `NewMemberAdded` reducer

File: src/routes/App/routes/new/modules/addBillForm/reducer.js [hunk 4]
```diff
@@ -1,23 +1,36 @@
...
+    let pseudoGlobalState = {
+      App: {
+        addBillForm: state
       }
-    ];
+    };
 
-    return stateCopy
-  };
+    let isTransientMemberFn = isTransientMemberFactory(pseudoGlobalState);
```
Comment (duffleit) [inline src/routes/App/routes/new/modules/addBillForm/reducer.js]: is it good practice to call selectors in the reducer? I thought the way to do it is `state -> reducer -> selector ` and not `state -> selector -> reducer`. The introduction of a `pseudoGlobalState` just to use a selector seems to be a bad tradeoff for me. Pretty hard to understand.

Comment (duffleit) [inline src/routes/App/routes/new/modules/addBillForm/reducer.js]: In addition, I think the comment is the best indicator that this section is hard to understand by someone who has to read this code :)

File: src/routes/App/routes/new/modules/addBillForm/selectors.test.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  const isExistingMember = isTransientMemberFactory(mockState());
+
+  expect(isExistingMember("Felix")).toBeTruthy();
+
+  expect(isExistingMember("Matthias")).toBeFalsy();
+  expect(isExistingMember("Thomas")).toBeFalsy();
+  expect(isExistingMember("David")).toBeFalsy();
+
+});
```
Comment (duffleit) [inline src/routes/App/routes/new/modules/addBillForm/selectors.test.js]: 👍 🥇

File: src/routes/App/routes/new/modules/addBillForm/reducer.js (hunk 1 above)
Comment (thomaseizinger) [inline src/routes/App/routes/new/modules/addBillForm/reducer.js]: Well, `isMemberInCurrentBill` is not completely correct. I explicitly want to check if the member is a __transient__ member, e.g. if they were added on the add-bill screen.

File: src/routes/App/routes/new/modules/addBillForm/reducer.js (hunk 2 above)
Comment (thomaseizinger) [inline src/routes/App/routes/new/modules/addBillForm/reducer.js]: Good point, will do.

File: src/routes/App/routes/new/modules/addBillForm/reducer.js (hunk 4 above)
Comment (thomaseizinger) [inline src/routes/App/routes/new/modules/addBillForm/reducer.js]: Well, depends on what you want to favor. I wanted to reuse the logic implemented in the selectors to check for transient members. IMO staying DRY is worth the introduction of the pseudo-state. After all, selectors are just functios.

File: src/routes/App/routes/new/modules/addBillForm/reducer.js (hunk 3 above)
Comment (thomaseizinger) [inline src/routes/App/routes/new/modules/addBillForm/reducer.js]: I was thinking about this. However, isn't it more readable to do all the construction of the new state in one place in comparison to have it scattered over several functions?

**Comment by thomaseizinger** [comment]
I have removed the complicated selector again and instead implemented a simple function inside the reducer. I also added some tests for the reducers and extracted action creators for the two actions I modified.

Check the commit messages for details :)

I have decided to keep the "duplication" of creating the PersonValues array in the NEW_MEMBER_ADDED and NEW_DEBTEE_SELECTED reducer because I think it is easier to read if all the spreading and overwriting of the previous state is in one place.

**Pull request merged**